### PR TITLE
Removing datastore.Dataset class and its uses.

### PR DESCRIPTION
**Pull request by dhermes**
Addresses part 8 of #477.

@tseaver What do you want to do about #499? We can finish that review and I can rebase this, but it will likely be quite difficult.

I have put off doc changes (unless necessary, e.g. the RST docs which would not build without the class) and will tackle that next.

I have also not yet been able to resolve the cyclic import in `Key.get()` but I am leaning towards simply removing it and renaming `datastore.__init__.get_entities` to `datastore.__init__.get` (and allowing a `Key` or a list of `Key`s).

### CODE REVIEW

**Comment by dhermes** [comment]
Also, I need to fix the following "extra" modules:

```
gcloud/datastore/demo/...
regression/populate_datastore.py
regression/clear_datastore.py
```

I have put it off for now since each can be done in an individual (easier to review) PR.

File: gcloud/datastore/key.py [hunk 1]
```diff
@@ -83,9 +83,8 @@ def _validate_dataset_id(self):
                  can be implied.
         """
         if self._dataset_id is None:
-            if _implicit_environ.DATASET is not None:
-                # This assumes DATASET.id() is not None.
-                self._dataset_id = _implicit_environ.DATASET.id()
+            if _implicit_environ.DATASET_ID is not None:
```
Comment (tseaver) [inline gcloud/datastore/key.py]: I have a sense that `_combine_args` and `_validate_dataset_id` should be combined.

Comment (dhermes) [inline gcloud/datastore/key.py]: SGTM. My main goal here was to erase all evidence of `Dataset` and not much else.

Mind handling in the cherry-pick of #499?

File: gcloud/datastore/test_helpers.py [hunk 2]
```diff
@@ -19,20 +19,18 @@ class Test_entity_from_protobuf(unittest2.TestCase):
...
         from gcloud.datastore import _implicit_environ
-        _implicit_environ.DATASET = self._replaced_dataset
         _implicit_environ.DATASET_ID = self._replaced_dataset_id
 
     def _callFUT(self, val):
         from gcloud.datastore.helpers import entity_from_protobuf
         return entity_from_protobuf(val)
 
-    def test_wo_dataset(self):
+    def test_wo_dataset_id(self):
```
Comment (tseaver) [inline gcloud/datastore/test_helpers.py]: We've lost something here (not this commit):  this test is identical to `test_w_dataset_id`

File: gcloud/datastore/__init__.py [hunk 3]
```diff
@@ -111,43 +104,16 @@ def get_connection():
...
-    :rtype: :class:`gcloud.datastore.dataset.Dataset`
-    :returns: A dataset with a connection using the provided credentials.
-    """
-    connection = get_connection()
-    return Dataset(dataset_id, connection=connection)
-
-
-def _require_dataset():
-    """Convenience method to ensure DATASET is set.
+def _require_dataset_id():
```
Comment (tseaver) [inline gcloud/datastore/__init__.py]: Cf ad78f59941219a7e5103873676fe6409a7b4a34f (passes the value in, tests explicitly against `None`).  Same for `_require_connection`.

File: gcloud/datastore/key.py (hunk 1 above)
Comment (tseaver) [inline gcloud/datastore/key.py]: Sure. that's fine.

File: gcloud/datastore/test_helpers.py (hunk 2 above)
Comment (dhermes) [inline gcloud/datastore/test_helpers.py]: Nice find, should I file a bug? Handle it here? You care to handle in cherry-pick of #499?

Comment (tseaver) [inline gcloud/datastore/test_helpers.py]: I'll pick it up w/ the cherrypick.

File: gcloud/datastore/__init__.py (hunk 3 above)
Comment (dhermes) [inline gcloud/datastore/__init__.py]: Want me to handle now or you want to? (FWIW I like the change.)

**Comment by dhermes** [comment]
@tseaver I am going to try to do a large docs fix-up. After your cherry-pick of #499 is in, how do you want to divvy up:

```
gcloud/datastore/demo/...
regression/populate_datastore.py
regression/clear_datastore.py
```

**Comment by tseaver** [comment]
how about I tackle the 'demo' stuff, and you the 'regression'?

**Comment by dhermes** [comment]
SGTM. I'm going to tackle docs first after we're done with #501 and then get to that.

**Pull request merged**